### ISSUE TITLE
Extending IdentityUserRole

### ISSUE THREAD

**Issue by RoySalisbury** [aspnet/Identity#956]
I have a need to extend the way roles (and claims) are assigned to a user.  I need to have the notion of a "site" that a role is applied to for a user.  For example, lets say that you have the following "Roles":

```
Administrator
Manager
User
```

Then lets say you have the following "sites"

```
Alabama
California
Texas
```

Normally you would say a user is a member of the "Manager" role and thats where it end.  However, I need to extend that to say your a "Manager" only for "Texas".  So I extended the IdentityUserRole class to include a SiteId.  (This has not foreign key as its a external ID).

```
 public class ApplicationUserRole<TKey> : IdentityUserRole<TKey> where TKey : IEquatable<TKey>
{
    public virtual int? SiteId { get; set;  }
}
```

And in my DbContext's OnModelCreating

```
builder.Entity<ApplicationUserRole<int>>(entity =>
{
    entity.HasKey(k => new { k.RoleId, k.SiteId, k.UserId });
});
```

So far its pretty straight forward.  The idea is that if the "SiteId" is null, then its basically like the default .. its a "global" role.  I have updated my UserManager/UserStore for some of the necessary updates and method additions to accommodate the new property.  All well and good.  I even had to create a new derived "UserClaimsPrincipalFactory<>" so that I created the necessary claim strings

```
e.g.:   claim/role: "<siteId>:<roleName>"
```

I got 90 % of it working and functioning as I would expect.  However, its the adding of new Roles that has issues.  It seems that adding a new primary key for the Db is fine, but it also creates a new unique constraint for { UserId, RoleId } (due to the foreign keys I expect). So I can not add new items.  If I go into the model and the DB and manually remove these constrains I can now add a new item.  But only one per DB context instance. If I do two in a row, then I get a key violation (its still only looking at the original key (I expect EF is silently adding it).

So with all that backstory .. How do you extend the IdentityUserRole?  Or is this a bug in EF?  Other than the adding, it seems to be working as I need it to (still have a lot of test cases to write).

**Comment by RoySalisbury** [comment]
I've worked around my issue (I hope), and will outline what I did below.  However, I think there needs to be something better.

Here is what I did.  

Do the base model creation from the parent.  Then, "ignored" the entire entity.

    base.OnModelCreating(builder);
    builder.Ignore<ApplicationUserRole<int>>();

However, that meant that some of the FK relationships were removed.  So I needed to add them back.

    builder.Entity<ApplicationUser>().HasMany(u => u.Roles).WithOne().HasForeignKey(ur => ur.UserId).IsRequired();
    builder.Entity<ApplicationRole>().HasMany(r => r.Users).WithOne().HasForeignKey(ur => ur.RoleId).IsRequired();

So, now I have the same model without the Key on IdentityUserRole from the base model.

Now, because the property I want to add is Nullable (int?), adding a complex key with all 3 fields marks my new field as not nullable in the DB.  So, I had to add a new PK field that I could use. So I added an Id field for the entity along with my own field.

    public class ApplicationUserRole<TKey> : IdentityUserRole<TKey> where TKey : IEquatable<TKey>
    {
        public virtual int Id { get; set;  }
        public virtual int? SiteId { get; set; }
    }

Next, I updated my model definition

   builder.Entity<ApplicationUserRole<int>>(entity =>
   {
       entity.HasKey(x => x.Id);
       entity.HasIndex(x => new { x.RoleId, x.UserId, x.SiteId }).IsUnique(true);
   });

Even adding an AlternateKey manually would mark the SiteId field as not nullable in the DB.  So I opted to use an index to enforce uniqueness. Once I updated my UserStore and UserManager classes to take into account the SiteId (optionally), it works. 

Perhaps if the entity (IdentityUserRole) had the PK as a unique Id and a AK or IX to enforce the uniqueness in the base class should be the way to go as the default. Then it is easier to extend.

**Comment by master-still** [comment]
@RoySalisbury 
Exactly same case ! Any Suggestions / Guidance ?

![identity_workaround](https://cloud.githubusercontent.com/assets/2611892/22171885/d4ae149e-dfc0-11e6-8529-51ea36be6d7a.png)

I must be missing something for my UserStore .

**Comment by RoySalisbury** [comment]
I'll post what I did to extend this tomorrow (don't have access to my code right now). Don't let me forget if you don't get a reply tomorrow. 

Roy

**Comment by master-still** [comment]
@RoySalisbury  The issue i mentioned above was solved !  [#988](https://github.com/aspnet/Identity/issues/1082#issuecomment-274238096)
